pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,239 @@
+}
+
+func (p *Plan) LastMemoryFootprint() int64 {
+	return p.lastNumBytes.Load()
+}
+
+// NewBlock returns a new block after allocating the time-range for fetch.
+func (p *Plan) NextBlock() (reference *Block, err error) {
+	timeDelta := determineTimeDelta(p.lastNumBytes.Load(), p.BlockSizeLimitBytes, p.deltaIncRegion, p.lastTimeRangeDelta)
+	mint := p.nextMint
+	maxt := mint + timeDelta
+	if maxt > p.Maxt {
+		maxt = p.Maxt
+	}
+	p.nextMint = maxt + 1
+	p.lastTimeRangeDelta = timeDelta
+	bRef, err := p.createBlock(mint, maxt)
+	if err != nil {
+		return nil, fmt.Errorf("next-block: %w", err)
+	}
+	return bRef, nil
+}
+
+// createBlock creates a new block and returns reference to the block for faster write and read operations.
+func (p *Plan) createBlock(mint, maxt int64) (reference *Block, err error) {
+	if err = p.validateT(mint, maxt); err != nil {
+		return nil, fmt.Errorf("create-block: %w", err)
+	}
+	id := p.blockCounts.Add(1)
+	timeRangeInMinutes := (maxt - mint) / time.Minute.Milliseconds()
+	percent := float64(maxt-p.Mint) * 100 / float64(p.Maxt-p.Mint)
+	if percent > 100 {
+		percent = 100
+	}
+	baseDescription := fmt.Sprintf("block-%d time-range: %d mins | mint: %d | maxt: %d", id, timeRangeInMinutes, mint, maxt)
+	reference = &Block{
+		id:                    id,
+		percent:               percent,
+		pbarDescriptionPrefix: baseDescription,
+		pbar: progressbar.NewOptions(
+			6,
+			progressbar.OptionOnCompletion(func() {
+				_, _ = fmt.Fprint(os.Stderr, "\n")
+			}),
+		),
+		mint:    mint,
+		maxt:    maxt,
+		pbarMux: p.pbarMux,
+		plan:    p,
+	}
+	if p.Quite {
+		reference.pbar = nil
+		reference.pbarDescriptionPrefix = ""
+	}
+	if p.TestCheckFunc != nil {
+		// This runs only during integration tests. It enables the tests to access and test the internal
+		// state of the planner.
+		p.TestCheckFunc()
+	}
+	reference.SetDescription(fmt.Sprintf("fetching time-range: %d mins...", timeRangeInMinutes), 1)
+	return
+}
+
+func (p *Plan) validateT(mint, maxt int64) error {
+	switch {
+	case p.Mint > mint || p.Maxt < mint:
+		return fmt.Errorf("invalid mint: %d: global-mint: %d and global-maxt: %d", mint, p.Mint, p.Maxt)
+	case p.Mint > maxt || p.Maxt < maxt:
+		return fmt.Errorf("invalid maxt: %d: global-mint: %d and global-maxt: %d", mint, p.Mint, p.Maxt)
+	case mint > maxt:
+		return fmt.Errorf("mint cannot be greater than maxt: mint: %d and maxt: %d", mint, maxt)
+	}
+	return nil
+}
+
+func determineTimeDelta(numBytes, limit, deltaIncRegion int64, prevTimeDelta int64) int64 {
+	switch {
+	case numBytes <= deltaIncRegion:
+		// We increase the time-range linearly for the next fetch if the current time-range fetch resulted in size that is
+		// less than half the aimed maximum size of the in-memory block. This continues till we reach the
+		// maximum time-range delta.
+		if prevTimeDelta >= maxTimeRangeDeltaLimit {
+			return maxTimeRangeDeltaLimit
+		}
+		return clampTimeDelta(prevTimeDelta, laIncrement)
```

Comment:
nit: see comment on clamp function
```suggestion
		return clampTimeDelta(prevTimeDelta + laIncrement)
```